### PR TITLE
Fix flash attention bugs with Mistral and Falcon

### PR DESCRIPTION
**Pull request by fxmarty**
This PR fixes some important bugs in the Mistral and Falcon integration.

https://github.com/huggingface/transformers/pull/26933 broke flash attention for Falcon due to the modification of the layout

The following tests were not passing:
```
FAILED tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_flash_attn_2_inference_padding_right - AssertionError: ValueError not raised

FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_flash_attn_2_generate_left_padding - RuntimeError: CUDA error: device-side assert triggered
FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_flash_attn_2_generate_padding_right - RuntimeError: CUDA error: device-side assert triggered
FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_flash_attn_2_generate_use_cache - RuntimeError: CUDA error: device-side assert triggered
FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_flash_attn_2_inference - RuntimeError: CUDA error: device-side assert triggered
FAILED tests/models/falcon/test_modeling_falcon.py::FalconModelTest::test_flash_attn_2_inference_padding_right - RuntimeError: CUDA error: device-side assert triggered
```

and Falcon with FA2 is not really usable on `main` due to an error in the shape (currently `[batch_size, num_head, seqlen, head_dim]` instead of the required `[batch_size, seqlen, num_head, head_dim]`.

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -838,7 +838,7 @@ def forward(
             attention_mask is not None
             and hasattr(self.config, "_flash_attn_2_enabled")
             and self.config._flash_attn_2_enabled
-            and past_key_values is not None
+            and use_cache
```
Comment (fxmarty) [inline src/transformers/models/mistral/modeling_mistral.py:841]: In the first autoregressive pass, `past_key_values` is None. In the following passes, we may have masks as

```
[[1, 1, 0, 0, 1],
[1, 1, 1, 1, 1]]
```

and `is_padding_right` is wrongfully evaluated to False.

File: src/transformers/models/falcon/modeling_falcon.py [hunk 2]
```diff
@@ -564,6 +564,12 @@ def forward(
 
         past_key_value = (key_layer, value_layer) if use_cache else None
 
+        # TODO: These transpose are quite inefficient but Flash Attention requires the layout [batch_size, sequence_length, num_heads, head_dim]. We would need to refactor the KV cache
```
Comment (younesbelkada) [inline src/transformers/models/falcon/modeling_falcon.py:567]: Thanks for fixing this!

**Review by younesbelkada — APPROVED** [review]
Clean ! Thanks !

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -2835,7 +2835,7 @@ def test_flash_attn_2_conversion(self):
 
         for model_class in self.all_model_classes:
             if not model_class._supports_flash_attn_2:
-                return
+                self.skipTest(f"{model_class.__name__} does not support Flash Attention 2")
```
Comment (amyeroberts) [inline tests/test_modeling_common.py:2838]: This is a lot nicer - thanks! 🤗

File: tests/models/llama/test_modeling_llama.py [hunk 4]
```diff
@@ -418,6 +421,50 @@ def test_flash_attn_2_generate_padding_right(self):
...
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_generate_use_cache(self):
+        import torch
+
+        max_new_tokens = 30
+
+        for model_class in self.all_generative_model_classes:
+            if not model_class._supports_flash_attn_2:
+                return
```
Comment (amyeroberts) [inline tests/models/llama/test_modeling_llama.py:435]:
```suggestion
                self.skipTest(f"{model_class.__name__} does not support Flash Attention 2")
```

File: tests/models/llama/test_modeling_llama.py [hunk 5]
```diff
@@ -385,6 +387,7 @@ def test_model_rope_scaling(self, scaling_type):
 
     @require_flash_attn
     @require_torch_gpu
+    @require_bitsandbytes
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:390]: good catch!

File: tests/models/llama/test_modeling_llama.py [hunk 6]
```diff
@@ -418,6 +421,50 @@ def test_flash_attn_2_generate_padding_right(self):
 
         self.assertListEqual(output_native, output_fa_2)
 
+    @require_flash_attn
+    @require_torch_gpu
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_generate_use_cache(self):
+        import torch
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:429]: isn't torch already imported?

File: tests/models/llama/test_modeling_llama.py [hunk 7]
```diff
@@ -418,6 +421,50 @@ def test_flash_attn_2_generate_padding_right(self):
...
+            if hasattr(config, "max_position_embeddings"):
+                config.max_position_embeddings = max_new_tokens + dummy_input.shape[1] + 1
+
+            model = model_class(config)
+
+            with tempfile.TemporaryDirectory() as tmpdirname:
+                model.save_pretrained(tmpdirname)
+
+                dummy_attention_mask = inputs_dict.get("attention_mask", torch.ones_like(dummy_input))
+                # NOTE: Mistral apparently does not support right padding + use_cache with FA2.
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:453]: it works but you'll get terrible results because the cache will cut the non padded values first

**Review by ArthurZucker — APPROVED** [review]
Thanks for fixing!

File: tests/models/mistral/test_modeling_mistral.py [hunk 8]
```diff
@@ -405,36 +405,52 @@ def test_flash_attn_2_generate_padding_right(self):
...
-    def test_flash_attn_2_inference_padding_right(self):
+    def test_flash_attn_2_generate_use_cache(self):
         import torch
 
-        for model_class in self.all_model_classes:
+        max_new_tokens = 30
+
+        for model_class in self.all_generative_model_classes:
             if not model_class._supports_flash_attn_2:
                 return
```
Comment (amyeroberts) [inline tests/models/mistral/test_modeling_mistral.py:415]:
```suggestion
                self.skipTest(f"{model_class.__name__} does not support Flash Attention 2")
```

File: tests/models/llama/test_modeling_llama.py (hunk 7 above)
Comment (amyeroberts) [inline tests/models/llama/test_modeling_llama.py:453]: Maybe?

```suggestion
                # NOTE: Llama apparently does not support right padding + use_cache with FA2.
```

File: tests/models/llama/test_modeling_llama.py [hunk 9]
```diff
@@ -418,6 +421,50 @@ def test_flash_attn_2_generate_padding_right(self):
...
+                    tmpdirname,
+                    torch_dtype=torch.float16,
+                    use_flash_attention_2=True,
+                    low_cpu_mem_usage=True,
+                ).to(torch_device)
+
+                # Just test that a large cache works as expected
+                _ = model.generate(
+                    dummy_input, attention_mask=dummy_attention_mask, max_new_tokens=max_new_tokens, do_sample=False
+                )
```
Comment (amyeroberts) [inline tests/models/llama/test_modeling_llama.py:466]: Just to make sure I've understood is `use_cache` set anywhere here? Is it necessary for these tests?

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

I just have a small Q about the tests

File: tests/models/llama/test_modeling_llama.py (hunk 6 above)
Comment (fxmarty) [reply tests/models/llama/test_modeling_llama.py:429]: good catch!

File: tests/models/llama/test_modeling_llama.py (hunk 4 above)
Comment (fxmarty) [reply tests/models/llama/test_modeling_llama.py:435]: adding test_flash_attn_2_generate_use_cache here was a mistake. Removed, good catch!

File: tests/models/llama/test_modeling_llama.py (hunk 7 above)
Comment (fxmarty) [reply tests/models/llama/test_modeling_llama.py:453]: adding test_flash_attn_2_generate_use_cache here was a mistake. Removed, good catch!

File: tests/models/llama/test_modeling_llama.py (hunk 9 above)
Comment (fxmarty) [reply tests/models/llama/test_modeling_llama.py:466]: adding test_flash_attn_2_generate_use_cache here was a mistake. Removed, good catch!

**Comment by fxmarty** [comment]
Hi @amyeroberts thank you for the review! I added by mistake `test_flash_attn_2_generate_use_cache` in `test_modeling_llama.py` while it was meant to be added in `test_modeling_mistral.py`, hence the confusion, apology!

**Comment by amyeroberts** [comment]
@fxmarty Thanks for clarifying. Tbh, I'm still a bit confused with the tests - it's not clear to me how this explicitly tests for the cache as `use_cache` isn't set anywhere 😅

**Comment by fxmarty** [comment]
@amyeroberts good catch indeed... I just checked, we are going here that sets `use_cache=True`: https://github.com/huggingface/transformers/blob/f93c1e9eceafde40b1d33fbb03834de97556706c/src/transformers/generation/utils.py#L1602

**Comment by amyeroberts** [comment]
@fxmarty OK - thanks for explaining. As a follow up, could you add `use_cache=True` explicitly into the tests? This way it's clearer for anyone who sees the code and isn't subject to silently not being tested anymore if the configs or config handling changes

**Comment by fxmarty** [comment]
For sure @amyeroberts I will ping you there. Sorry I should have waited before merging..

**Comment by amyeroberts** [comment]
@fxmarty No worries! It doesn't affect the functionality of this PR so it's fine to be done separately :)

**Pull request merged**